### PR TITLE
Wait for request body if necessary in filter

### PR DESCRIPTION
**Pull request by yawkat**
Use FullHttpRequest.bufferContents to wait for the request body if it is needed for a filter method but isn't immediately available.

Also changed the bufferContents contract a bit so that this will fail more gracefully in servlet.

### CODE REVIEW

**Comment by yawkat** [comment]
@dstepanov I think it would be really useful to have an ArgumentBinder that returns an ExecutionFlow<BindingResult<T>> at some point, it would save us a lot of trouble

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import io.micronaut.http.annotation.ServerFilter
+import io.micronaut.http.client.HttpClient
+import io.micronaut.runtime.server.EmbeddedServer
+import jakarta.inject.Singleton
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Sinks
+import spock.lang.Specification
+
+class FilterBodySpec extends Specification {
+    def "delayed body"() {
```
Comment (graemerocher) [inline http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy]: can you add another test that tests the case where a controller method is reached after a request filter processes the delayed body

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy (hunk 1 above)
Comment (yawkat) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy]: ok

**Review by graemerocher — APPROVED** [review]

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        @RequestFilter("/**")
+        @Order(-1)
+        void signal() {
+            secondBodyPartSink.emitValue("bar", Sinks.EmitFailureHandler.FAIL_FAST)
+        }
+
+        @RequestFilter("/delayed-body")
+        @Order(0)
+        HttpResponse<String> await(@Body String body) {
```
Comment (dstepanov) [inline http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy:66]: Why would you have two filters with the same order?

**Review by dstepanov — APPROVED** [review]

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy (hunk 2 above)
Comment (yawkat) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy:66]: one for testing with the controller and one for without

**Comment by dstepanov** [comment]
I think we should add an option to wait for the body to the `@RequestFilter` for use cases when you want to inject simple `HttpRequest` and read the body.

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy (hunk 2 above)
Comment (dstepanov) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/filters/FilterBodySpec.groovy:66]: Yeah I didn't notice the different path

**Comment by graemerocher** [comment]
yes perhaps if the request specifies a concrete type like `HttpRequest<Foo>`

**Pull request merged**